### PR TITLE
Test deploy

### LINKED ISSUE #2: Add unit tests for contract deployment.
We currently have no test coverage for contract deployment. A new test should be added (there is an `unimplemented!()` placeholder in `src/contract/deploy.rs`) that goes through a deployment. Since deployment uses the `TransactionBuilder::execute_and_confirm` API we will get test coverage on that method as well. There are a couple deployment paths that need to be tested, the paths that require testing being:
- Deploy contract without libraries and without constructor
- Deploy contract with constructor without parameters
- Deploy contract with constructor that requires parameters
- Deploy contract with library(ies)
- Deploy contract with signed transaction (either with `Account::Locked` or `Account::Offline`)
- Deploy contract with transaction request (with `Account::Local`)

Note that some of these paths, but not all are mutually exclusive so we probably don't need a test for each of these to achieve full coverage.

For some context on how to set up the test transport for these take a look at the tests in `src/transaction.rs`.

### PR DESCRIPTION
**Pull request by e00E**
fixes #2

### CODE REVIEW

**Comment by e00E** [comment]
@nlordell Is this roughly what you had in mind? Asking before I add more tests.

File: src/contract/deploy.rs [hunk 1]
```diff
@@ -264,4 +258,63 @@ mod tests {
...
+            .unwrap();
+
+        transport.assert_request("eth_accounts", &[]);
+        transport.assert_request(
+            "eth_sendTransaction",
+            &[json!({"data": bytecode, "from": account})],
+        );
+        transport.assert_request("eth_blockNumber", &[]);
+        transport.assert_request("eth_getTransactionReceipt", &[json!(transaction_hash)]);
+        transport.assert_no_more_requests();
```
Comment (nlordell) [inline src/contract/deploy.rs:318]: Maybe we can assert that the `Instance` result address is matches what was returned from the tx receipt.

**Review by nlordell — COMMENTED** [review]
That was exactly what I had in mind. What other `deploy` tests did you have in mind?

File: src/contract/deploy.rs [hunk 2]
```diff
@@ -264,4 +258,63 @@ mod tests {
         assert_eq!(error.to_string(), DeployError::EmptyBytecode.to_string());
         transport.assert_no_more_requests();
     }
+
+    #[test]
+    fn deploy() {
+        let mut transport = TestTransport::new();
+        let web3 = Web3::new(transport.clone());
+
+        let mut artifact = Artifact::empty();
```
Comment (nlordell) [inline src/contract/deploy.rs:267]: nit:
```
let artifact = Artifact {
    bytecode: Bytecode::from_hex_str(bytecode).unwrap(),
    ..Artifact::empty()
};
```

File: src/contract/deploy.rs [hunk 3]
```diff
@@ -264,4 +258,63 @@ mod tests {
...
+        transport.add_response(json!(transaction_hash));
+
+        // eth_blockNumber
+        transport.add_response(json!("0x0"));
+
+        // eth_getTransactionReceipt
+        let contract_address = addr!("0x9876543210987654321098765432109876543210");
+        let block_hash =
+            hash!("0x4242424242424242424242424242424242424242424242424242424242424242");
+        let mut logs_bloom = "0x".to_string();
```
Comment (nlordell) [inline src/contract/deploy.rs:288]:
```suggestion
        let logs_bloom = H2048::repeat_byte(0);
```
https://docs.rs/web3/0.10.0/web3/types/struct.H2048.html#method.repeat_byte

Comment (nlordell) [reply src/contract/deploy.rs:288]: This can probably be fixed in other locations in the code.

**Comment by e00E** [comment]
> That was exactly what I had in mind. What other `deploy` tests did you have in mind?

The ones listed in #2 .

**Comment by nlordell** [comment]
@e00E Oh yeah, I forgot about my list. Sure, note that I think there might be a bug with transaction signing for transactions that deploy contracts.

Take a look here for how RLP encoding should work for transaction where the `to` address is `None`:
https://github.com/tomusdrw/rust-web3/blob/ab762e0f48c527a19952c3eb776fc7a7fb14f52e/src/api/accounts.rs#L267-L312

Opened issue #181

**Pull request closed without merging**